src/main/java/io/burt/athena/result/s3/AsyncResumableGetObjectOperation.java

```diff
@@ -0,0 +1,222 @@
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Consumer;
+
+
+public class AsyncResumableGetObjectOperation<ReturnT> implements Callable<CompletableFuture<ReturnT>> {
+    private final S3AsyncClient s3AsyncClient;
+    private final GetObjectRequest.Builder requestBuilder;
+    private final AsyncResponseTransformer<GetObjectResponse, ReturnT> targetTransformer;
+    private int retryLimit;
+
+    public AsyncResumableGetObjectOperation(S3AsyncClient s3AsyncClient, Consumer<GetObjectRequest.Builder> request, AsyncResponseTransformer<GetObjectResponse, ReturnT> targetTransformer, int retryLimit) {
+        this(s3AsyncClient, GetObjectRequest.builder().applyMutation(request), targetTransformer, retryLimit);
+    }
+
+    public AsyncResumableGetObjectOperation(S3AsyncClient s3AsyncClient, GetObjectRequest request, AsyncResponseTransformer<GetObjectResponse, ReturnT> targetTransformer, int retryLimit) {
+        this(s3AsyncClient, request.toBuilder(), targetTransformer, retryLimit);
+    }
+
+    public AsyncResumableGetObjectOperation(S3AsyncClient s3AsyncClient, GetObjectRequest.Builder requestBuilder, AsyncResponseTransformer<GetObjectResponse, ReturnT> targetTransformer, int retryLimit) {
+        this.s3AsyncClient = s3AsyncClient;
+        this.requestBuilder = requestBuilder;
+        this.targetTransformer = targetTransformer;
+        this.retryLimit = retryLimit;
+    }
+
+    public CompletableFuture<ReturnT> call() {
+        return s3AsyncClient.getObject(
+            requestBuilder.build(),
+            new WrapperAsyncResponseTransformer()
+        );
+    }
+
+    private class WrapperAsyncResponseTransformer implements AsyncResponseTransformer<GetObjectResponse, ReturnT>, SdkPublisher<ByteBuffer>, Subscription, Subscriber<ByteBuffer> {
+        private final AtomicReference<SdkPublisher<ByteBuffer>> sourcePublisher = new AtomicReference<>();
+        private final AtomicReference<Subscription> sourceSubscription = new AtomicReference<>();
+        private final AtomicBoolean isBeforeFirstBody = new AtomicBoolean(true);
+        private final AtomicBoolean isCancelled = new AtomicBoolean();
+        private final AtomicReference<Subscriber<? super ByteBuffer>> targetSubscriber = new AtomicReference<>();
+        private final Set<Throwable> delayedExceptions = new CopyOnWriteArraySet<>();
+        private final AtomicBoolean hasSubscribed = new AtomicBoolean();
+        private final AtomicBoolean hasOnSubscribedTarget = new AtomicBoolean();
+        private final LongAdder bytesReceived = new LongAdder();
```

Comment:
There are most probably a bit many atomics here. Taking the `targetSubscriber` example, the reason it is atomic is for the `failTarget` callback that could be called concurrently by multiple threads, but we should provide the guarantee to the target subscriber that we call `onError` (or `onComplete`) exactly once. This could obviously be implemented using synchronized or an explicit lock instead, which would probably make the code more clear as the synchronization is only needed in one location. That being said, we don't need more than an atomic in this case, and it feels easier to reason about the lack of deadlocks with only an atomic.

The synchronized blocks around the request counters is because they are affected both by the producer and the consumer, and requires multiple variables to be kept in sync, so it was fairly complicated to use atomic to solve that.

But there are likely multiple mistakes in this aspect of the implementation, as the ART contract is a very complicated one, and the fact that both the consumer and producer can act concurrently doesn't help.